cni/pkg/install/install.go

```diff
@@ -47,42 +46,44 @@ func NewInstaller(cfg *config.InstallConfig, isReady *atomic.Value) *Installer {
 	}
 }
 
-func (in *Installer) install(ctx context.Context) error {
-	if err := copyBinaries(in.cfg.CNIBinSourceDir, in.cfg.CNIBinTargetDirs); err != nil {
+func (in *Installer) install(ctx context.Context) ([]string, error) {
+	copiedFiles, err := copyBinaries(in.cfg.CNIBinSourceDir, in.cfg.CNIBinTargetDirs)
+	if err != nil {
 		cniInstalls.With(resultLabel.Value(resultCopyBinariesFailure)).Increment()
-		return fmt.Errorf("copy binaries: %v", err)
+		return copiedFiles, fmt.Errorf("copy binaries: %v", err)
 	}
 
 	if err := writeKubeConfigFile(in.cfg); err != nil {
 		cniInstalls.With(resultLabel.Value(resultCreateKubeConfigFailure)).Increment()
-		return fmt.Errorf("write kubeconfig: %v", err)
+		return copiedFiles, fmt.Errorf("write kubeconfig: %v", err)
 	}
 
 	cfgPath, err := createCNIConfigFile(ctx, in.cfg)
 	if err != nil {
 		cniInstalls.With(resultLabel.Value(resultCreateCNIConfigFailure)).Increment()
-		return fmt.Errorf("create CNI config file: %v", err)
+		return copiedFiles, fmt.Errorf("create CNI config file: %v", err)
 	}
 	in.cniConfigFilepath = cfgPath
 
-	return nil
+	return copiedFiles, nil
 }
 
 // Run starts the installation process, verifies the configuration, then sleeps.
 // If an invalid configuration is detected, the installation process will restart to restore a valid state.
 func (in *Installer) Run(ctx context.Context) error {
-	if err := in.install(ctx); err != nil {
+	installedBins, err := in.install(ctx)
+	if err != nil {
 		return err
 	}
 	installLog.Info("Installation succeed, start watching for re-installation.")
 
 	for {
-		if err := in.sleepCheckInstall(ctx); err != nil {
+		if err := in.sleepCheckInstall(ctx, installedBins); err != nil {
 			return err
 		}
 
 		installLog.Info("Detect changes to the CNI configuration and binaries, attempt reinstalling...")
-		if err := in.install(ctx); err != nil {
+		if _, err := in.install(ctx); err != nil {
```

Comment:
shouldn't we update `installedBins` now?

TBH I am not sure its really going to change, but in theory there could be more things added to the source dir

Reply:
Yeah I thought about it - I left it like this because everything we actually care to watch _has_ to be there from the beginning/first install, and if it's not that's explicitly a bug/we shouldn't absorb it on subsequent reinstalls - a "reinstall" shouldn't silently pull in more source binaries.